### PR TITLE
MonadParsec instances for StateT and RWST

### PR DESCRIPTION
**Pull request by DanielRrr**
The full implementation with some util

### CODE REVIEW

File: Megaparsec/Megaparsec.lean [hunk 1]
```diff
@@ -33,19 +22,65 @@ structure PosState (S: Type) where
...
+    match u with
+    | .tokens nelᵤ => match v with
+                      | .tokens nelᵥ => ord2beq_nel T nelᵤ nelᵥ
+                      | _ => false
+    | .label nelᵤ => match v with
+                     | .label nelᵥ => ord2beq_nel Char nelᵤ nelᵥ
+                     | _ => false
+    | .eof => match v with
+              | .eof => true
+              | _ => false
```
Comment (arthurpaulino) [inline Megaparsec/Megaparsec.lean:55]:
```suggestion
def ord2beq_nel (T: Type) [Ord T] (x y: NonEmptyList T): Bool :=
  match x, y with
  | .cons u x₁, .cons v y₁ => ord2beq T u v && ord2beq_nel T x₁ y₁
  | .nil u, .nil v => ord2beq T u v
  | _, _ => false

instance ord2beq_ei [Ord T]: BEq (ErrorItem T) where
  beq (u v: ErrorItem T) :=
    match u, v with
    | .tokens nelᵤ, .tokens nelᵥ => ord2beq_nel T nelᵤ nelᵥ
    | .label nelᵤ, .label nelᵥ => ord2beq_nel Char nelᵤ nelᵥ
    | .eof, .eof => true
    | _, _ => false
```
I think it's cleaner this way

File: Megaparsec/Megaparsec.lean [hunk 2]
```diff
@@ -33,19 +22,65 @@ structure PosState (S: Type) where
...
+
 inductive ErrorFancy (E: Type) where
 | fail (msg: String)
 | indent (ord: Ordering) (fromPos: Pos) (uptoPos: Pos)
 | custom (e: E)
 
-inductive ParseError (S E: Type) [Stream S] where
-| trivial (offset: Nat) (unexpected: (ErrorItem (Stream.Token S))) (expected: List (ErrorItem (Stream.Token S)))
+inductive ParseError (S E: Type) [stream: Stream S] where
+| trivial (offset: Nat) (unexpected: (@ErrorItem (Stream.Token S))) (expected: List (@ErrorItem (Stream.Token S)))
```
Comment (arthurpaulino) [inline Megaparsec/Megaparsec.lean:83]:
```suggestion
inductive ParseError (S E: Type) [Stream S] where
| trivial (offset: Nat) (unexpected: ErrorItem (Stream.Token S)) (expected: List (ErrorItem (Stream.Token S)))
```

File: Megaparsec/Megaparsec.lean [hunk 3]
```diff
@@ -60,33 +95,198 @@ inductive Result (S E A: Type) [Stream S] where
...
+-- #eval (4 : Id Nat) + (4 : Nat)
+#eval (4 : Id Nat)
+#eval (4 : Nat)
+#eval (pure 4 : Id Nat)
+#eval ((pure 4 : Id Nat) : Nat)
+
+--instance: Monad (MonadParsec M) where
+
+--instance: MonadLiftT (MonadParsec Mᵤ) Mᵥ where
+--monadLift {α: Type} (???):
```
Comment (arthurpaulino) [inline Megaparsec/Megaparsec.lean:198]:
```suggestion
```

File: Megaparsec/Megaparsec.lean [hunk 4]
```diff
@@ -60,33 +95,198 @@ inductive Result (S E A: Type) [Stream S] where
...
+def pMap [Stream S] [Monad M] (f: U → V) (x: ParsecT E S M U): ParsecT E S M V :=
   ParsecT.mk (λ (b s cok cerr eok eerr) => (x.unParser b s (cok ∘ f) cerr (eok ∘ f) eerr))
 
+instance [Stream S] [Monad M]: Functor (ParsecT E S M) where
+  map := pMap
+
+-- def pSeq [Stream S] [Monad M] (step: ParsecT E S M (U -> V)) (start: Unit -> ParsecT E S M U): ParsecT E S M V :=
+--   let cok₁ := sorry
+--   let eok₁ := sorry
+--   ParsecT.mk (λ b s cok cerr eok eerr => step.unParser b s cok₁ cerr eok₁ eerr)
```
Comment (arthurpaulino) [inline Megaparsec/Megaparsec.lean:150]:
```suggestion
```

**Review by arthurpaulino — COMMENTED** [review]
Most of the changes are nitpick. But I see a lot of comments that seem to serve as documentation. We should use proper docstrings for those:
```lean
/-- this is a docstring -/
```

File: Megaparsec/Megaparsec.lean [hunk 5]
```diff
@@ -169,58 +166,127 @@ class MonadParsec (M: Type -> Type) [Monad M] [Alternative M] where
...
+  | ⟨ a, _ ⟩ => ⟨ a, c ⟩
+
+def fst {A B : Type} : A × B → A
+  | ⟨ a, _⟩ => a 
+
+def seqComp {A B : Type} {M : Type → Type} [Monad M] (ma : M A) (mb : M B) :=
+  ma >>= λ _ => mb
+--
+
+instance (E S : Type) [Monad M] [Alternative M] [Stream S] [mₚ: MonadParsec E S M] : MonadParsec E S (StateT σ M) where
```
[the remark this answers is not in the document]
Comment (cognivore) [reply Megaparsec/Megaparsec.lean:196]: I originally added it just as a MonadPlus

File: Megaparsec/Megaparsec.lean [hunk 6]
```diff
@@ -169,58 +152,117 @@ class MonadParsec (M: Type -> Type) [Monad M] [Alternative M] where
...
+
+-- | RWS monad and its transformer and 
+
+def void {A : Type} {M : Type → Type} [Monad M] (ma : M A) : M Unit :=
+  (λ _ => Unit.unit) <$> ma
+
+def RWST (R W S : Type u) (M : Type u → Type v) (A : Type u) : Type (max u v) :=
+  R → S → M (A × S × W)
+
+instance (R W S : Type) (M : Type → Type) [m : Monoid W]  [Monad M] : Monad (RWST R W S M) where
```
Comment (cognivore) [inline Megaparsec/Megaparsec.lean:221]: Are you sure it'll automatically derive Pure, Bind, Functor etc? I like Lean's granular type classes, but maybe it also automatically derive stuff up the class hierarchy.

**Review by cognivore — APPROVED** [review]

**Pull request merged**